File: hogvm/typescript/src/stl/crypto.ts
import { ExecOptions } from '../types'

export function sha256Hex(data: string, options?: ExecOptions): string {
    const crypto = options?.external?.crypto
    if (!crypto) {
        throw new Error('The crypto module is required for "sha256Hex" to work.')
    }
    return crypto.createHash('sha256').update(data).digest('hex')
}

export function md5Hex(data: string, options?: ExecOptions): string {
    const crypto = options?.external?.crypto
    if (!crypto) {
        throw new Error('The crypto module is required for "md5Hex" to work.')
    }
    return crypto.createHash('md5').update(data).digest('hex')
}

export function sha256HmacChainHex(data: string[], options?: ExecOptions): string {
    const crypto = options?.external?.crypto
    if (!crypto) {
        throw new Error('The crypto module is required for "sha256HmacChainHex" to work.')
    }
    if (data.length < 2) {
        throw new Error('Data array must contain at least two elements.')
    }
    let hmac = crypto.createHmac('sha256', data[0])
    hmac.update(data[1])
    for (let i = 2; i < data.length; i++) {
        hmac = crypto.createHmac('sha256', hmac.digest())
        hmac.update(data[i])
    }
    return hmac.digest('hex')
}

File: hogvm/python/stl/crypto.py
import hashlib
import hmac


def md5Hex(data: str) -> str:
    return hashlib.md5(data.encode()).hexdigest()


def sha256Hex(data: str) -> str:
    return hashlib.sha256(data.encode()).hexdigest()


def sha256HmacChainHex(data: list) -> str:
    if len(data) < 2:
        raise ValueError("Data array must contain at least two elements.")

    hmac_obj = hmac.new(data[0].encode(), data[1].encode(), hashlib.sha256)
    for i in range(2, len(data)):
        hmac_obj = hmac.new(hmac_obj.digest(), data[i].encode(), hashlib.sha256)

    return hmac_obj.hexdigest()
